Android/app/src/test/java/app/intra/net/doh/google/GoogleServerConnectionIntegrationTest.java

```diff
@@ -73,41 +88,63 @@ public void testPerformDnsRequest() throws Exception {
         verify(mockDb, times(1)).setPreferred(any(DualStackResult.class));
     }
 
+    @Test
+    public void testPerformDnsRequestFallback() throws Exception {
+        when(mockDb.lookup(GoogleServerConnection.PRIMARY_TLS_HOSTNAME)).thenReturn(Arrays.asList(InetAddress.getAllByName
+          (GoogleServerConnection.HTTP_HOSTNAME)));
+        when(mockDb.lookup(GoogleServerConnection.FALLBACK_TLS_HOSTNAME)).thenReturn(Arrays.asList(InetAddress.getAllByName
+          (GoogleServerConnection.HTTP_HOSTNAME)));
+        GoogleServerConnection s = GoogleServerConnection.get(mockDb, new PrimaryHostnameFailureInterceptor());
+
+        verify(mockDb, times(1)).lookup(GoogleServerConnection.PRIMARY_TLS_HOSTNAME);
```

Comment:
These verify calls are testing the bootstrap fallback.  It would be good to test after the performDnsRequest call below that only FALLBACK_TLS_HOSTNAME is being looked up.

We might also add another test where the bootstrap doesn't require fallback, the first performDnsRequest fails on PRIMARY_TLS_HOSTNAME and needs to fallback to FALLBACK_TLS_HOSTNAME, and subsequent performDnsRequest calls only use FALLBACK_TLS_HOSTNAME.

Reply:
Sounds good, I'll add your suggestions in a follow-up PR.